app/models/service_order.rb

```diff
@@ -90,6 +90,24 @@ def self.order_immediately(request, requester)
                         :tenant       => requester.current_tenant).checkout_immediately
   end
 
+  def deep_copy(new_attributes = {})
+    dup.tap do |new_service_order|
+      # Set it to nil - the after_create hook will give it the correct name
+      new_service_order.name = nil
+      # Should be put back into the Cart state
+      new_service_order.state = STATE_CART
+      new_service_order.miq_requests = miq_requests.collect do |request|
+        request.class.send(:create, request.attributes.except("id", "name", "approval_state", "request_state",
+                                                              "message", "created_on", "fulfilled_on", "updated_on",
+                                                              "placed_on"))
+      end
+      new_attributes.each do |attr, value|
+        new_service_order.send("#{attr}=", value)
```

Comment:
This worries me a bit - maybe because we're implicitly trusting what is coming in to the method `new_attributes = {}`.  I understand the `save!` would validate and throw any exception.  Maybe I'm being a bit too cautious.

@gmcculloug do you have any thoughts?